rfcs/0004-transaction.md

```diff
@@ -0,0 +1,288 @@
+
+At the moment of writing, SlateDB still haven't have the equivant of sequence number yet, instead, we could regard a manifest + wal number to represent a consistent view of a table.
+
+There're an in-progress RFC on exploring the design on Snapshot. We can refer to that RFC for the details later. However, we'd **assume that each key has a sequence number** in this doc, because it's considered as a common practice in LSM-tree based storage engines.
+
+#### Write Batch
+
+This is a sample [code snippet](https://rust.velas.com/rocksdb/struct.WriteBatch.html) of `WriteBatch` in the `rocksdb` crate:
+
+```
+use rocksdb::{DB, Options, WriteBatch};
+
+let path = "_path_for_rocksdb_storage1";
+{
+    let db = DB::open_default(path).unwrap();
+    let mut batch = WriteBatch::default();
+    batch.put(b"my key", b"my value");
+    batch.put(b"key2", b"value2");
+    batch.put(b"key3", b"value3");
+    db.write(batch); // Atomically commits the batch
+}
+```
+
+You may notice that the `WriteBatch` API is very similar to the `Transaction` API we proposed above, the write operation from `WriteBatch` is also considered as atomic. The only differences are:
+
+1. the `WriteBatch` is committed by `db.write(batch)`, while the `Transaction` is committed by `txn.commit()`
+2. beside the write operations, the `Transaction` should also support the read operations, and if one key is updated in the `Transaction`, the read operation should return the updated value during the transaction.
+
+The `WriteBatch` is considered as a prerequisite for the transaction feature, as it plays as a place to buffer the writes, and it makes MVCC possible with an iterator over it in front of the iterator over Snapshot. So we should implement the `WriteBatch` first before the transaction feature.
+
+We can refer to the WriteBatch RFC or PR for the details later.
+
+## Conflict Checking: Snapshot Isolation
+
+The conflict checking in the SI transaction is simple:
+
+1. Track the sequence number when transaction is started in the `TransactionState` struct.
+2. Track the modified keys in the `TransactionState` struct.
+3. When `commit()` is called, check whether the modified keys are modified by others after the transaction started.
+
+The `TransactionState` struct could be like:
+
+```rust
+struct TransactionState {
+  started_seq: u64,
+  write_keys: HashSet<String>,
+}
+```
+
+How to verify the modified keys are modified by others after the transaction started?
+
+We could simply check the sequence number of the keys in latest db storage. Given a key `"key1"`, let's say the sequence number of the current transaction is "102", while the sequence number of the key `"key1"` is "103" in the storage, then we could regard the key `"key1"` is modified by others after the transaction started, thus violates the conflict check and should abort the commit.
```

Comment:
yes, in my understanding, the conflict check & writing the write batch should be serialized together. it might not be an "real" lock if we had serialize all the writes behind a single commit pipeline, as no other thread will contend with the writes.

Reply:
Ah, right! I forgot about the commit pipeline. 😀